src/layoutBuilder.js

```diff
@@ -388,75 +495,224 @@ function convertWordsToCodepoints(lineAsArrayOfCodepoints) {
+
+	wordsAsArraysOfCodePoints.forEach((wordAsCodePoints) => {
+		const wordAsString = spreadify(
+			String.fromCharCode,
+			String
+		)(wordAsCodePoints);
+		const bidiWord = bidi
+			.from_string(wordAsString, { direction: 'RTL' })
+			.reorder_visually()
+			.toString();
+
+		if (bidiWord !== ' ') {
+			let originalTextnode = wordPropsLookup[bidiWord.trim()] || {};
+			const newTextNode = Object.assign({}, originalTextnode, {
+				text: bidiWord,
+			});
+			arrayOfTransformedWords.push(newTextNode);
+		}
+	});
+
+	// need to add a space manually to the end of an RTL inline run
+	arrayOfTransformedWords[arrayOfTransformedWords.length - 1].text += ' ';
+
+	const updatedInlines = textTools.buildInlines(
+		[{ text: arrayOfTransformedWords, style: node.style }],
+		styleStack
+	);
+
+	return updatedInlines.items;
+}
+
+/**
+ * Takes the inlines from the current line. Reorders the inlines using BIDI. After reordering
+ * the inlines it uses them to make a new line and returns that line
+ * @param {Line} line The original line that contains some inline RTL
+ * @param {StyleContextStack} styleStack The style context used when rebuilding the inlines
+ * @param {TextTools} textTools Used to rebuild the inlines
+ * @param {Object} textNode The original textNode that contains the nested inline rtl text
+ * @param {Number} availableWidth the line width being used, required to build a new line
+ * @returns {Line}
+ */
+
+function addLineWithInlineRTL(line, styleStack, textTools, textNode, availableWidth) {
+	// line.inlines contains all the words that fit on the line. We will create a new line and
+	// loop through these words. If the word is not RTL, we'll add it to the new line immediately.
+	// If the line is RTL, we'll collect up all the RTL words in that block and then transform
+	// them before adding them to the new line.
+	// Finally, return the new line.
+
+	const inlines = line.inlines;
+	const newLine = new Line(availableWidth);
+
+	while (inlines.length) {
+		const currentInline = inlines.shift();
+		if (!currentInline.inlineRtl) {
+			newLine.addInline(currentInline);
+		} else {
+			// if we're dealing with rtl, we need to slice out a chunk and transform it
+			// before we add it to the new line
+			const nextLTRIndex = inlines.findIndex((i) => !i.inlineRtl);
+			let rtlInlines = [currentInline];
+			if (nextLTRIndex !== -1) {
+				// in this case we have an RTL block that ends before the line end
+				rtlInlines = rtlInlines.concat(inlines.splice(0, nextLTRIndex));
+			} else {
+				// here we have an RTL block running all the way to the end of the line
+				rtlInlines = rtlInlines.concat(inlines.splice(0));
 			}
-			return line.addInline(newInline);
-		});
+			const transformedInlines = transformRtlInlines(
+				rtlInlines,
+				textNode,
+				styleStack,
+				textTools
+			);
+			transformedInlines.forEach((transformedInline) => {
+				newLine.addInline(transformedInline);
+			});
+		}
+	}
+
+	return newLine;
```

Comment:
Second new function. This is hopefully a bit easier to understand than the one above.

Give it a line, it'll make a new line then add each word one by one _unless_ it's an RTL word. If it's an RTL word, it'll grab the run of RTL words, send them off for processing, then add the processed (ie correctly ordered) words to the line.

Once all the words are added, it returns the new line.